data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriter.java

```diff
@@ -429,26 +534,51 @@ private static Converter buildConverter(TypeDescription schema) {
       case SHORT:
         return new ShortConverter();
       case DATE:
+        return new DateConverter();
       case INT:
         return new IntConverter();
       case LONG:
-        return new LongConverter();
+        ORCSchemaUtil.LongType longType = ORCSchemaUtil.LongType.valueOf(
+            schema.getAttributeValue(ORCSchemaUtil.ICEBERG_LONG_TYPE_ATTRIBUTE));
+        switch (longType) {
+          case TIME:
+            return new TimeConverter();
+          case LONG:
+            return new LongConverter();
+          default:
+            throw new IllegalStateException("Unhandled Long type found in ORC type attribute: " + longType);
+        }
       case FLOAT:
         return new FloatConverter();
       case DOUBLE:
         return new DoubleConverter();
       case BINARY:
-        return new BytesConverter();
+        ORCSchemaUtil.BinaryType binaryType = ORCSchemaUtil.BinaryType.valueOf(
+            schema.getAttributeValue(ORCSchemaUtil.ICEBERG_BINARY_TYPE_ATTRIBUTE));
+        switch (binaryType) {
+          case UUID:
+            return new UUIDConverter();
+          case FIXED:
+            return new FixedConverter();
+          case BINARY:
+            return new BytesConverter();
+          default:
+            throw new IllegalStateException("Unhandled Binary type found in ORC type attribute: " + binaryType);
+        }
       case STRING:
       case CHAR:
       case VARCHAR:
         return new StringConverter();
       case DECIMAL:
-        return schema.getPrecision() <= 18 ?
-            new Decimal18Converter(schema) :
-            new Decimal38Converter(schema);
+        // TODO: Figure out the right cases in which to use the fastpath Decimal18Converter
```

Comment:
Should we create an issue to track this?

Reply:
I just ended up fixing the issue in the latest commit. So should no longer be required.